### PR TITLE
Cleanup articles with duplicate user_id, title and body_markdown

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR contains two scripts

1. the first one deletes duplicate draft articles that have the same combination of `user_id`, `title` and `body_markdown`. The algorithm is the same of https://github.com/forem/forem/pull/10138 and https://github.com/forem/forem/pull/10239

1. the second one "merges" published articles:
  - basically we divide the articles in groups (some articles have 1 duplicate, some have more)
  - of all the groups we take the first article
  - we copy over the data from the others onto this first one
  - we destroy the others

This is where the data is:

- https://dev.to/admin/blazer/queries/232-duplicate-draft-articles-body_markdown-user_id-title
- https://dev.to/admin/blazer/queries/235-duplicate-published-articles-body_markdown-user_id-title
- https://dev.to/admin/blazer/queries/236-published-articles-of-users-with-duplicates

## QA Instructions, Screenshots, Recordings

Ideally this should be tested on a copy of the production DB. To make sure it worked I downloaded the sample of duplicates from https://dev.to/admin/blazer/queries/236-published-articles-of-users-with-duplicates - imported them locally and ran the scripts to make sure they worked. 

If you want to test it:

- download a CSV from https://dev.to/admin/blazer/queries/236-published-articles-of-users-with-duplicates
- open the SQL console with `psql PracticalDeveloper_development`
- type `\d articles`
- note the names of the foreign key constraints
- drop the foreign key constraints on the `articles` table, otherwise the `user_id` and `organization_id` won't work:

```sql
alter table articles drop constraint fk_rails_3d31dad1cc;
alter table articles drop constraint fk_rails_7809a1a57d;
```

remember that the constraint names might be different than mine. Please double check with `\d articles` in the SQL console and replace them with yours
- import the data from the downloaded CSV:

```sql
copy articles(id,user_id,title,body_html,slug,created_at,updated_at,main_image,description,published,published_at,featured,processed_html,featured
 _number,social_image,body_markdown,organization_id,canonical_url,collection_id,show_comments,main_image_background_hex_color,receive_notifications,approved,published_from_feed,c
 omments_count,reactions_count,video,video_code,video_source_url,video_thumbnail_url,video_closed_caption_track_url,hotness_score,feed_source_url,second_user_id,third_user_id,las
 t_buffered,positive_reactions_count,edited_at,crossposted_at,spaminess_rating,language,facebook_last_buffered,cached_tag_list,path,cached_user_name,cached_user_username,last_com
 ment_at,boost_states,video_state,page_views_count,previous_positive_reactions_count,score,reading_time,comment_template,video_duration_in_seconds,experience_level_rating,experie
 nce_level_rating_distribution,last_experience_level_rating_at,rating_votes_count,originally_published_at,organic_page_views_count,organic_page_views_past_month_count,organic_pag
 e_views_past_week_count,cached_user,cached_organization,archived,any_comments_hidden,nth_published_by_author,comment_score,search_optimized_title_preamble,search_optimized_descr
 iption_replacement,public_reactions_count,previous_public_reactions_count,user_subscriptions_count) from '/path/to/file.csv.csv' DELIMITER ',' csv header;
```

don't forget to replace the path of the file with the full path of the file you downloaded

- count the number of articles, just to have a quick idea if the script has worked
- replace `.destroy_all` with `.delete_all` on line https://github.com/forem/forem/pull/10273/files#diff-ae11923370b79ce32bb9abbeebca964dR43 (the reason for this is that you won't have the same user objects locally, but we need to pass through Rails to make sure everything left attached to articles is actually destroyed)
- run the data update script with `rails data_updates:run` if it's the first time or manually in the console

```ruby
require Rails.root.join("lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb")
DataUpdateScripts::CleanupPublishedArticlesWithDuplicateUserIdTitleBodyMarkdown.new.run
```

- re-count the number of articles, it should be less ;-)
- run https://dev.to/admin/blazer/queries/235-duplicate-published-articles-body_markdown-user_id-title locally to make sure they are gone
- also run the two group by queries - https://dev.to/admin/blazer/queries/81-duplicate-draft-articles-by-body_markdown-user_id-title and https://dev.to/admin/blazer/queries/227-duplicate-published-articles-by-body_markdown-user_id-title to double check

### CODE REVIEW

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
+module DataUpdateScripts
+  class CleanupPublishedArticlesWithDuplicateUserIdTitleBodyMarkdown
```
Comment (rhymes) [inline lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:2]: In an internal discussion on Slack we've decided to go ahead and "merge" duplicate articles

cc @mstruve

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    private
+
+    # Main goal here is copy and merge everything from the duplicates to the main one
+    # NOTE: I'm intentionally ignore the case that two duplicates belong to different orgs
+    # or different collections as it's highly unlikely
+    def graft_articles(article_to_keep, articles_to_graft)
+      article_id = article_to_keep.id
+      articles_to_graft_ids = articles_to_graft.map { |a| a["id"] }
+
+      models_with_a_direct_relation = [
```
Comment (rhymes) [inline lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:65]: I went through `schema.rb` and found those tables with `article_id` in them

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        # Poll is ignored because it's related to the liquid tag inside the article,
+        # also user's can't use polls
+        RatingVote,
+        # TagAdjustment is ignored as there's likely no reason for article to have an adjustment moved over
+      ]
+      models_with_a_direct_relation.each do |model_class|
+        model_class.where(article_id: articles_to_graft_ids).update_all(article_id: article_id)
+      end
+
+      Comment.where(commentable_type: "Article",
```
Comment (rhymes) [inline lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:80]: these are the tables that are polymorphic, I think we have all the meaningful ones

**Comment by mstruve** [comment]
Forking production then will run all of this fun stuff against it :)

File: lib/data_update_scripts/20200910135958_remove_draft_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+module DataUpdateScripts
+  class RemoveDraftArticlesWithDuplicateUserIdTitleBodyMarkdown
+    def run
+      # Currently there are duplicate draft articles in the DB with the same body_markdown, title, user_id combination
+
+      # This statement deletes all draft articles in excess found to be duplicate over those three columns
+      ActiveRecord::Base.connection.execute(
```
Comment (mstruve) [inline lib/data_update_scripts/20200910135958_remove_draft_articles_with_duplicate_user_id_title_body_markdown.rb:7]: This statement works perfectly!

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        expires_in: 2.weeks,
+      )
+    end
+
+    private
+
+    # Main goal here is copy and merge everything from the duplicates to the main one
+    # NOTE: I'm intentionally ignore the case that two duplicates belong to different orgs
+    # or different collections as it's highly unlikely
+    def graft_articles(article_to_keep, articles_to_graft)
```
Comment (mstruve) [inline lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:61]: This transaction block has a few issues. 
1) Duplicate Notification Subscriptions when we try to roll them over `duplicate key value violates unique constraint "idx_notification_subs_on_user_id_notifiable_type_notifiable_id"` 
2) Duplicate Rating votes when we try to roll them over `duplicate key value violates unique constraint "index_rating_votes_on_user_id_and_article_id_and_context"`
3) Duplicate Notifications when we try to roll them over `duplicate key value violates unique constraint "index_notifications_on_user_notifiable_and_action_not_null"` - Honestly, I think we don't roll these over and we let them be destroyed. 

I tried to fix the first couple by doing this within the transaction block. 
```ruby
  NotificationSubscription.where(
    notifiable_type: "Article",
    notifiable_id: articles_to_graft_ids
  ).each do |notification_sub|
    notification_sub.update(notifiable_id: article_id)
  rescue ActiveRecord::RecordNotUnique
  end
```
However, that does not work bc as it turns out if Postgres raises an exception, it poisons future transactions on the same connection and you end up with this error
```
ActiveRecord::StatementInvalid (PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block)
```

TL;DR We need a way to avoid duplicates when we are rolling over records and my 6pm brain is not able to figure it out right now ;)

**Review by mstruve — COMMENTED** [review]
Couple things to iron out!

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 5 above)
Comment (citizen428) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:61]: > if Postgres raises an exception, it poisons future transactions on the same connection

This is a setting called `ON_ERROR_ROLLBACK`.

When it's off statements after the error will not be executed:

```
PracticalDeveloper_development# begin;
BEGIN
PracticalDeveloper_development*# select 1/0;
ERROR:  division by zero
PracticalDeveloper_development!# select 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
PracticalDeveloper_development!# rollback;
ROLLBACK
```

When it's on, statements continue executing (this implicitly uses [`ROLLBACK TO SAVEPOINT`](https://www.postgresql.org/docs/12/sql-rollback-to.html)):

PracticalDeveloper_development# \set ON_ERROR_ROLLBACK
PracticalDeveloper_development# begin;
BEGIN
PracticalDeveloper_development*# select 1/0;
ERROR:  division by zero
PracticalDeveloper_development*# select 1;
┌──────────┐
│ ?column? │
╞══════════╡
│        1 │
└──────────┘
(1 row)

PracticalDeveloper_development*# commit;
COMMIT
```

That said, option 3 seems sensible to me.

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                      ROW_NUMBER() OVER(PARTITION BY user_id, title, body_markdown) AS row_number,
+                      COUNT(*) OVER(PARTITION BY user_id, title, body_markdown) AS num_rows_in_partition
+               FROM articles) duplicates
+          WHERE published = 't'
+              AND num_rows_in_partition > 1 -- we select all rows in each partition with duplicates
+        SQL
+      )
+
+      articles_to_delete_ids = []
+      ActiveRecord::Base.transaction do
```
Comment (citizen428) [inline lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:19]: The default transaction isolation level is read committed. I assume that's what we want here, but double checking doesn't hurt.

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      models_with_a_direct_relation = [
+        BufferUpdate,
+        HtmlVariantSuccess,
+        HtmlVariantSuccess,
+        HtmlVariantTrial,
+        PageView,
+        # Poll is ignored because it's related to the liquid tag inside the article,
+        # also user's can't use polls
+        RatingVote,
+        # TagAdjustment is ignored as there's likely no reason for article to have an adjustment moved over
```
Comment (citizen428) [inline lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:74]: Will this potentially leave tag adjsutments for articles that no longer exist? If yes, should we clean them up?

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 5 above)
Comment (rhymes) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:61]: Thanks a lot @mstruve for the test! So glad unique indexes in the DB are working, we wouldn't have never caught these duplicates if we'd only have the Rails layer :D All this work we've been doing is validated by these errors.

I feel we have 2 options here: 

1. use a similar technique as with tags (check if they already exists, if they don't add them, otherwise skip)
1. ignore these three models and forget about it

I'll set a time slot to see if I can come up with a simple logic, otherwise we move to plan B, though being able to merge those as well should be helpful if we decide to move this script behind an admin page in the end

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 7 above)
Comment (rhymes) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:74]: They are going to be deleted when we call `.destroy_all` afterwards. They are just not migrated.

To be 100% nitpicky we should probably migrate those adjustments belonging to tags that the original article didn't have as we merge the tags but I'm not really sure it's worth it

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 5 above)
Comment (rhymes) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:61]: Decided to ignore NotificationSubscription, Notification and RatingVote :)

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 6 above)
Comment (rhymes) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:19]: Would you suggest otherwise? I'm not sure we need a different one

**Comment by rhymes** [comment]
Added the PR to add the index https://github.com/forem/forem/pull/10284

**Review by mstruve — APPROVED** [review]
Ran the new script for merging articles and it worked like a charm! 20 articles were removed. I then ran the blazer SQL query again to verify they were all gone and they were.

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 7 above)
Comment (citizen428) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb]: Agreed @rhymes, it's not worth it.

File: lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb (hunk 6 above)
Comment (citizen428) [reply lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb:19]: No, I think it's fine. I just wanted to double check, all good 👍

**Review by citizen428 — APPROVED** [review]
